impeller/core/host_buffer.cc

```diff
@@ -30,104 +39,101 @@ void HostBuffer::SetLabel(std::string label) {
 BufferView HostBuffer::Emplace(const void* buffer,
                                size_t length,
                                size_t align) {
-  auto [device_buffer, range] = state_->Emplace(buffer, length, align);
+  auto [data, range, device_buffer] = state_->Emplace(buffer, length, align);
   if (!device_buffer) {
     return {};
   }
-  return BufferView{state_, device_buffer, range};
+  return BufferView{std::move(device_buffer), data, range};
 }
 
 BufferView HostBuffer::Emplace(const void* buffer, size_t length) {
-  auto [device_buffer, range] = state_->Emplace(buffer, length);
+  auto [data, range, device_buffer] = state_->Emplace(buffer, length);
   if (!device_buffer) {
     return {};
   }
-  return BufferView{state_, device_buffer, range};
+  return BufferView{std::move(device_buffer), data, range};
 }
 
 BufferView HostBuffer::Emplace(size_t length,
                                size_t align,
                                const EmplaceProc& cb) {
-  auto [buffer, range] = state_->Emplace(length, align, cb);
-  if (!buffer) {
+  auto [data, range, device_buffer] = state_->Emplace(length, align, cb);
+  if (!device_buffer) {
     return {};
   }
-  return BufferView{state_, buffer, range};
+  return BufferView{std::move(device_buffer), data, range};
 }
 
 std::shared_ptr<const DeviceBuffer> HostBuffer::GetDeviceBuffer(
     Allocator& allocator) const {
-  return state_->GetDeviceBuffer(allocator);
+  return nullptr;
 }
 
 void HostBuffer::Reset() {
   state_->Reset();
 }
 
-size_t HostBuffer::GetSize() const {
-  return state_->GetReservedLength();
-}
-
-size_t HostBuffer::GetLength() const {
-  return state_->GetLength();
+void HostBuffer::HostBufferState::MaybeCreateNewBuffer(size_t required_size) {
+  if (current_buffer + 1 >= device_buffers.size()) {
+    if (required_size > kAllocatorBlockSize) {
+      FML_LOG(ERROR) << "Created oversized buffer: " << required_size;
```

Comment:
This can be addressed later when we start doing `DeviceBuffer` reuse, but it would probably be a good to track oversized `DeviceBuffer`s separately from the block list. Otherwise one oversized draw could end up causing a few oversized blocked to get allocated while scrolling around, for example.

In the doc for https://github.com/flutter/flutter/issues/138161 I recommended just throwing away oversized buffers and not tracking when starting out... But perhaps later we could form a heap for tracking oversize buffers sorted by size.

Reply:
Done